File: packages/whisker-status-bar/ios/Sources/WhiskerStatusBar/StatusBarModule.swift
// `whisker-status-bar` ModuleDefinition (iOS).
//
// A view-less DSL module: `definition()` has no `View(...)` block, just
// module-level `Function`s. The SwiftPM codegen plugin discovers the
// `Module` subclass, emits a `@_cdecl` dispatch shim, and registers it
// so `whisker::platform_module::invoke("WhiskerStatusBar", ...)` from
// Rust routes into these handlers.
//
// iOS decides status-bar appearance per view controller by default, so
// these app-level `UIApplication` setters are no-ops UNLESS
// `UIViewControllerBasedStatusBarAppearance = false` is set in
// Info.plist — which this crate's plugin injects (see `src/plugin.rs`).
// The setters are deprecated (since iOS 9) but remain fully functional
// with that plist key, and are the only way for a view-less module to
// drive the status bar without owning a view controller. This mirrors
// what `expo-status-bar` does on iOS.
//
// All calls hop to the main thread — UIKit status-bar mutation is main-
// thread-only, and a Rust caller may invoke from any thread.

import UIKit
import WhiskerModule

public final class StatusBarModule: Module {
    public override func definition() -> ModuleDefinition {
        ModuleDefinition {
            Name("WhiskerStatusBar")

            // setHidden(hidden: Bool) -> Null
            Function("setHidden") { (args: [WhiskerValue]) -> WhiskerValue in
                let hidden = args.first?.asBool ?? false
                DispatchQueue.main.async {
                    UIApplication.shared.setStatusBarHidden(hidden, with: .fade)
                }
                return .null
            }

            // setStyle(style: "light" | "dark") -> Null
            Function("setStyle") { (args: [WhiskerValue]) -> WhiskerValue in
                let style: UIStatusBarStyle =
                    (args.first?.asString == "light") ? .lightContent : .darkContent
                DispatchQueue.main.async {
                    UIApplication.shared.setStatusBarStyle(style, animated: true)
                }
                return .null
            }
        }
    }
}

File: packages/whisker-status-bar/android/src/main/kotlin/rs/whisker/modules/status_bar/StatusBarModule.kt
// `whisker-status-bar` ModuleDefinition (Android).
//
// A view-less DSL module: `definition()` has no `View(...)` block, just
// module-level `Function`s. The KSP processor finds the `Module`
// subclass and registers its functions with `WhiskerModuleRegistry`
// under the `Name(...)`, so `whisker::platform_module::invoke(
// "WhiskerStatusBar", ...)` from Rust routes into these handlers.
//
// Status-bar visibility + icon style are driven through
// `WindowInsetsControllerCompat` (androidx.core), which backports the
// modern `WindowInsetsController` API down to API 21. All mutation runs
// on the UI thread (`activity.runOnUiThread`) since window changes are
// main-thread-only and a Rust caller may invoke from any thread.

package rs.whisker.modules.status_bar

import android.app.Activity
import androidx.core.view.WindowCompat
import androidx.core.view.WindowInsetsCompat
import androidx.core.view.WindowInsetsControllerCompat
import rs.whisker.runtime.Module
import rs.whisker.runtime.ModuleDefinition
import rs.whisker.runtime.WhiskerValue

public class StatusBarModule : Module() {
    public override fun definition(): ModuleDefinition = ModuleDefinition {
        Name("WhiskerStatusBar")

        // setHidden(hidden: Bool) -> Null | Err
        Function("setHidden") { args ->
            try {
                val hidden = args.getOrNull(0)?.asBool() ?: false
                onActivity { activity ->
                    StatusBar.setHidden(activity, hidden)
                }
                WhiskerValue.Null
            } catch (t: Throwable) {
                WhiskerValue.Err("WhiskerStatusBar.setHidden failed: ${t.message}")
            }
        }

        // setStyle(style: "light" | "dark") -> Null | Err
        Function("setStyle") { args ->
            try {
                val style = args.getOrNull(0)?.asString() ?: "dark"
                onActivity { activity ->
                    StatusBar.setStyle(activity, style)
                }
                WhiskerValue.Null
            } catch (t: Throwable) {
                WhiskerValue.Err("WhiskerStatusBar.setStyle failed: ${t.message}")
            }
        }
    }

    /// Resolve the host activity and run `block` on its UI thread. No-op
    /// if no activity is attached (e.g. dispatch during teardown).
    private fun onActivity(block: (Activity) -> Unit) {
        val activity = appContext.currentActivity ?: return
        activity.runOnUiThread { block(activity) }
    }
}

/// Plain helper — no Whisker types. Kept separate from `StatusBarModule`
/// so the window logic is testable/readable on its own, matching the
/// `whisker-haptics` `HapticsModule`/`Haptics` split.
private object StatusBar {
    private fun controller(activity: Activity): WindowInsetsControllerCompat =
        WindowInsetsControllerCompat(activity.window, activity.window.decorView)

    fun setHidden(activity: Activity, hidden: Boolean) {
        val controller = controller(activity)
        if (hidden) {
            controller.systemBarsBehavior =
                WindowInsetsControllerCompat.BEHAVIOR_SHOW_TRANSIENT_BARS_BY_SWIPE
            controller.hide(WindowInsetsCompat.Type.statusBars())
        } else {
            controller.show(WindowInsetsCompat.Type.statusBars())
        }
    }

    fun setStyle(activity: Activity, style: String) {
        // `isAppearanceLightStatusBars = true` → dark icons (for a light
        // background); expo's `style="dark"`. `false` → light/white icons
        // (for a dark background); expo's `style="light"`.
        controller(activity).isAppearanceLightStatusBars = (style == "dark")
    }
}
